src/lsf/man/man1/lsmon.1
.ds ]W %
.ds ]L
.nh
.TH lsmon 1 "Lava Version 1.0 - Sept 2007"
.br
.SH NAME
\fBlsmon\fR - displays load information for Lava hosts and periodically updates the 
display
.SH SYNOPSIS
.BR
.PP
.PP
\fBlsmon\fR [\fB-N\fR\fI \fR| \fB-E\fR] [\fB-n\fR \fInum_hosts\fR] [\fB-R\fR \fIres_req\fR] [\fB-I\fR \fIindex_list\fR] [\fB-i\fR \fIinterval\fR] 
[\fB-L\fR \fIfile_name\fR] [\fIhost_name\fR ...]
.PP
\fBlsmon\fR [\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRlsmon is a full-screen Lava monitoring utility that displays and updates 
load information for hosts in a cluster. 
.PP
By default, displays load information for all hosts in the cluster, up to 
the number of lines that will fit on-screen. 
.PP
By default, displays raw load indices.
.PP
By default, load information is sorted according to CPU and paging 
load. 
.PP
By default, load information is updated every 10 seconds.
.SH OPTIONS
.BR
.PP
.TP 
\fB-N
\fR
.IP
Displays normalized CPU run queue length load indices (see 
lsfintro(1)). 


.TP 
\fB-E
\fR
.IP
Displays effective CPU run queue length load indices (see 
lsfintro(1)). Options -N and -E are mutually exclusive. 


.TP 
\fB-n\fR \fInum_hosts
\fR
.IP
Displays only load information for the requested number of hosts. 
Information for up to \fInum_hosts\fR hosts that best satisfy resource 
requirements is displayed. 


.TP 
\fB-R\fR\fI \fR\fIres_req\fR 

.IP
Displays only load information for hosts that satisfy the specified 
resource requirements. See lsinfo(1) for a list of built-in resource 
names. 

.IP
Load information for the hosts is sorted according to load on the 
specified resources. 

.IP
If \fIres_req\fR contains special resource names, only load information for 
hosts that provide these resources is displayed (see lshosts(1) to find 
out what resources are available on each host). 

.IP
If one or more host names are specified, only load information for the 
hosts that satisfy the resource requirements is displayed. 


.TP 
\fB-I\fR \fIindex_list
\fR
.IP
Displays only load information for the specified load indices. Load 
index names must be separated by a colon (for example, r1m:pg:ut). 

.IP
If the index list \fIindex_list\fR is too long to fit in the screen of the user 
who invoked the command, the output is truncated. For example, if the 
invoker's screen is 80 characters wide, then up to 10 load indices are 
displayed.


.TP 
\fB-i\fR \fIinterval
\fR
.IP
Sets how often load information is updated on-screen, in seconds.


.TP 
\fB-L\fR \fIfile_name
\fR
.IP
Saves load information in the specified file while it is displayed on-
screen. 

.IP
If you do not want load information to be displayed on your screen at 
the same time, use \fBlsmon -L\fR \fIfile_name\fR \fB< /dev/null\fR. The format 
of the file is described in lim.acct(5).


.TP 
\fIhost_name\fR ...

.IP
Displays only load information for the specified hosts. 


.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits. 


.TP 
\fB-V\fR 

.IP
Prints Lava release version to stderr and exits.


.SH USAGE
.BR
.PP
.PP
You can use the following commands while lsmon is running:
.PP
[\fB^L\fR |\fB i\fR | \fBn\fR | \fBN\fR | \fBE\fR | \fBR\fR | \fBq\fR]
.TP 
\fB^L
\fR
.IP
Refreshes the screen.


.TP 
\fBi
\fR
.IP
Prompts you to input a new update interval.


.TP 
\fBn
\fR
.IP
Prompts you to input a new number of hosts to display.


.TP 
\fBN
\fR
.IP
Toggles between displaying raw CPU run queue length load indices 
and normalized CPU run queue length load indices.


.TP 
\fBE
\fR
.IP
Toggles between displaying raw CPU run queue length load indices 
and effective CPU run queue length load indices.


.TP 
\fBR
\fR
.IP
Prompts you to input new resource requirements.


.TP 
\fBq
\fR
.IP
Quits lsmon.


.SH OUTPUT
.BR
.PP
.PP
The following fields are displayed by default. 

.IP
HOST_NAME
.BR
.RS
.IP
Name of specified hosts for which load information is displayed, or 
if resource requirements were specified, name of hosts that satisfied 
the specified resource requirement and for which load information 
is displayed. 

.RE

.IP
status
.BR
.RS
.IP
Status of the host. A minus sign (-) may precede the status, 
indicating that the Remote Execution Server (RES) on the host is not 
running. 

.IP
Possible statuses are: 


.IP
ok 
.BR
.RS
.IP
The host is in normal load sharing state and can accept remote 
jobs. 

.RE

.IP
busy
.BR
.RS
.IP
The host is overloaded because some load indices exceed 
configured thresholds. Load index values that caused the host 
to be busy are preceded by an asterisk (*). Built-in load indices 
include r15s, r1m, r15m, ut, pg, io, ls, it, swp, mem and tmp 
(see below). External load indices are configured in the file 
lsf.cluster.\fIcluster_name\fR (see lsf.cluster(5)). 

.RE

.IP
lockW 
.BR
.RS
.IP
The host is locked by its run window. Run windows for a host 
are specified in the configuration file (see lsf.conf(5)) and 
can be displayed by lshosts. A locked host will not accept 
load shared jobs from other hosts. 

.RE

.IP
lockU 
.BR
.RS
.IP
The host is locked by the Lava administrator or root. 

.RE

.IP
unavail 
.BR
.RS
.IP
The host is down or the Load Information Manager (LIM) on 
the host is not running. 

.RE

.IP
.RE

.IP
r15s
.BR
.RS
.IP
The 15-second exponentially averaged CPU run queue length.

.RE

.IP
r1m
.BR
.RS
.IP
The 1-minute exponentially averaged CPU run queue length.

.RE

.IP
r15m
.BR
.RS
.IP
The 15-minute exponentially averaged CPU run queue length.

.RE

.IP
ut
.BR
.RS
.IP
The CPU utilization exponentially averaged over the last minute, 
between 0 and 1.

.RE

.IP
pg
.BR
.RS
.IP
The memory paging rate exponentially averaged over the last 
minute, in pages per second.

.RE

.IP
ls
.BR
.RS
.IP
The number of current login users.

.RE

.IP
it
.BR
.RS
.IP
On UNIX, the idle time of the host (keyboard not touched on all 
logged in sessions), in minutes.
.RE

.IP
tmp
.BR
.RS
.IP
The amount of free space in /tmp, in megabytes.

.RE

.IP
swp
.BR
.RS
.IP
The amount of currently available swap space, in megabytes.

.RE

.IP
mem
.BR
.RS
.IP
The amount of currently available memory, in megabytes.

.RE
.SH SEE ALSO
.BR
.PP
.PP
lsfintro(1), lshosts(1), lsinfo(1), lsload(1), 
lslockhost(8), lim.acct(5), ls_load(3)
.SH DIAGNOSTICS
.BR
.PP
.PP
Specifying an invalid resource requirement string while lsmon is 
running (via the R option) causes lsmon to exit with an appropriate 
error message.
.PP
lsmon exits if it does not receive a reply from LIM within the update 
interval.


src/lsf/lstools/lsmon.c
/* $Id: lsmon.c,v 1.4 2007/08/15 22:18:55 tmizan Exp $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */
#include "lsf/lib/liblavalite.h"







extern void (*Signal_ (int, void (*)(int)))(int);


extern int makeFields(struct hostLoad *, char *loadval[], char **);
extern void *formatHeader(char **, char);
extern char **filterToNames(char *);
                                             
void  printHeader(char **);

#define CONTROL_L  12  
#define MAXLISTSIZE 256

char    resreq[MAXLINELEN] = "";
char    indexfilter[MAXLINELEN] = "";
char    *hostnames[MAXLISTSIZE];
int     numneeded, num = 0, options = 0;
int     updateRate = 10;               
FILE    *lf;			
int     Lflag = 0;		
int     redirect = 0;           
char    logFilename[MAXFILENAMELEN];
extern int num_loadindex;


#define NL_SETN  27 


void
usage(char * cmd)
{
	
    fprintf(stderr, 
            "%s: %s [-h] [-V] [-N|-E] [-n num_hosts] [-R res_req] [-I index_list] [-i interval] [-L logfile] [host_name ... ]\n",
	    I18N_Usage,cmd); 
    exit(0);
}



void
endPgm(char * message, int status)
{
    
    if (!redirect) {
        echo();
        nocbreak();
        move(LINES-1, 0);
        refresh();
        endwin(); 
        putchar('\n');
    } 

    if (message != NULL) {
       ls_perror(message);
    }
    if (Lflag)
	fclose(lf);
    exit(status);
}

void
quit()
{
    endPgm(NULL, 0);
}

void
getLoadInfo(void)
{
    int  i, j, nf;  
    static int  numStatus;  
    struct hostLoad *hosts;
    int  numwanted;
    static char **loadval = NULL;
    static char first=true;
    static char *defaultindex[]={"r15s", "r1m", "r15m", "ut", "pg", "ls",
				  "it", "tmp", "swp", "mem", NULL};
    char **nlp;
    char linebuf[132];

    numwanted = numneeded;
    if (num > 0)
       options = options | EXACT;

    if (indexfilter[0] != '\0')
        nlp = filterToNames(indexfilter);
    else
        nlp = defaultindex;

    hosts = ls_loadinfo(resreq, &numwanted, options, 0, hostnames, num,
                             &nlp);      

    if (!hosts) {
	numwanted=0;
	if (!redirect) {
            mvprintw(3, 0, ls_sysmsg());
            clrtobot();
            move(1,0);
            refresh();
        }
        return;
    }

    printHeader(nlp);
    
    if (!redirect)
        move(3,0);

    if (loadval == NULL)
        if (!(loadval=(char **)malloc(num_loadindex*sizeof(char *)))) {
            lserrno=LSE_MALLOC;
            ls_perror("lsmon");
            exit(-1);
        }
   
    if (!redirect) 
        for (i = 0 ; (i < numwanted) && (i < (LINES - 4)) ; i++) {
	    

 	    mvprintw(i + 3, 0, "%-15.15s ", hosts[i].hostName);

	    
	    if (LS_ISUNAVAIL(hosts[i].status)) 
	        addstr(I18N_unavail);
	    else if (!LS_ISUNAVAIL(hosts[i].status) && LS_ISRESDOWN(hosts[i].status)) {
		      char str[7];

		      sprintf(str, "-%2s    ", I18N_ok);
		      addstr(str);
		  }
	    else if (LS_ISOKNRES(hosts[i].status)) {
		      char str[7];

		      sprintf(str, " %2s    ", I18N_ok);
		      addstr(str);
		  }
	    else if (LS_ISLOCKED(hosts[i].status))
	        {
	 	    addstr((_i18n_msg_get(ls_catd,NL_SETN,2107, "lock"))); /* catgets  2107  */
		    if (LS_ISLOCKEDU(hosts[i].status))
		        addch('U');
		    if (LS_ISLOCKEDW(hosts[i].status))
		        addch('W');
		    if (LS_ISLOCKEDM(hosts[i].status))
		        addch('M');
		    addstr("   ");
	        }
            else if (LS_ISBUSY(hosts[i].status))
		    addstr((_i18n_msg_get(ls_catd,NL_SETN,2108, "busy   "))); /* catgets  2108  */

	    if (!LS_ISUNAVAIL(hosts[i].status)) {
	        nf = makeFields(&hosts[i], loadval, nlp);
                strcpy(linebuf, loadval[0]);
                for(j=1; j < nf; j++) 
                    strcat(linebuf, loadval[j]);

               mvprintw(i + 3, 22, "%s",linebuf);
	    }

	if (Lflag) {
            if (first) {
	       numStatus = 0;
               for(j=0; nlp[j]; j++) {
                   fprintf(lf,"%s ",nlp[j]); 
		   numStatus++;
               }
               fprintf(lf,"\n");
               first=false;
            }

	    fprintf(lf, "%ld %s", time(NULL), hosts[i].hostName);

            for (j = 0; j < 1 + GET_INTNUM (numStatus); j++) {
		if (j  == GET_INTNUM (numStatus))
		    fprintf(lf, " %d ", hosts[i].status[j]);
                else
		    fprintf(lf, " %d", hosts[i].status[j]);
            }
	    for (j=0; nlp[j]; j++)
		fprintf(lf, "%f ", hosts[i].li[j]);
	    fprintf(lf, "\n");
	}
	    clrtoeol();
    } 
   else { 
       for (i=0; i < numwanted; i++) {
	    if (Lflag) {
		if (first) {
		    numStatus = 0;
 	            for(j=0; nlp[j]; j++) {
		        fprintf(lf,"%s ",nlp[j]);
			numStatus++;
                    }
		    fprintf(lf,"\n");
		    first=false;
		 }
		fprintf(lf, "%ld %s", time(NULL), hosts[i].hostName);
		for (j = 0; j < 1 + GET_INTNUM (numStatus); j++) {
		    if (j  == GET_INTNUM (numStatus))
		        fprintf(lf, " %d ", hosts[i].status[j]);
                    else
		        fprintf(lf, " %d", hosts[i].status[j]);
                }
		for (j=0; nlp[j]; j++)
		    fprintf(lf, "%f ", hosts[i].li[j]);
	        fprintf(lf, "\n");
	     }
       }
    } 
  
    if (!redirect) {
        if (i < (LINES - 4))
        {
            move(i+3, 0);
            clrtobot();
        }
        move(1,0);
        refresh();
    }
}


void
printHeader(char **nlp)
{
    char *hname;
    char *headLine;
    
    if (!redirect)
        mvaddstr(0, 0, _i18n_msg_get(ls_catd,NL_SETN,2109, "Hostname: ")); /* catgets  2109  */
    if ((hname = ls_getmyhostname()) == NULL) {
	endPgm("ls_getmyhostname", -1);
    }

    
    headLine = formatHeader(nlp, false);

    if (!redirect) {
        addstr(hname);
       mvprintw(0, COLS - 25, 
	   _i18n_msg_get(ls_catd,NL_SETN,2110, "Refresh rate: %3d secs"), /* catgets  2110  */
	   updateRate); 
		mvaddstr(2, 0, headLine);
    }

}


static void
chInterval(void)
{
   
   int        interval = 0;
   char       str[255];

   


   mvaddstr(1, 0, _i18n_msg_get(ls_catd,NL_SETN,2111, "Enter a new refresh rate (secs):  ")); /* catgets  2111  */
   clrtoeol();
   refresh();

   echo();
   getstr(str);
   noecho();

   interval = atoi(str);

   if (interval <= 0) {
      mvaddstr(1, 0, _i18n_msg_get(ls_catd,NL_SETN,2112, "Refresh rate must be greater than zero.  Interval unchanged.")); /* catgets  2112  */
   } else {
      updateRate = interval;
   }

   clrtoeol();
   refresh();
}


int
chNumNeeded(void)
{
   
   int        numhosts;
   char       str[255];

   
   move(1,0);
   clrtoeol();
   mvaddstr(1, 0, _i18n_msg_get(ls_catd,NL_SETN,2113, "Enter number of hosts to be displayed:  ")); /* catgets  2113  */
   refresh();
   echo();
   getstr(str);
   noecho();
   numhosts = atoi(str);

   if (numhosts <= 0) {
      move(1,0);
      clrtoeol();
      mvaddstr(1, 0, _i18n_msg_get(ls_catd,NL_SETN,2114, "Number of hosts must be greater than zero. Number unchanged.")); /* catgets  2114  */
      numhosts = numneeded; 
   } else {
      move(1,0);
      clrtoeol();
   }

   move(1,0);
   refresh();

   return numhosts;
}

void
chResReq(void)
{
   
   char       newresreq[MAXLINELEN];

   
   move(1,0);
   clrtoeol();
   mvaddstr(1, 0, _i18n_msg_get(ls_catd,NL_SETN,2115, "Enter resource requirements string:  ")); /* catgets  2115  */
   refresh();

   echo();
   getstr(newresreq);
   noecho();

   
   if (expSyntax_(newresreq) < 0) {
      
      mvaddstr(1, 0, 
           _i18n_msg_get(ls_catd,NL_SETN,2116, "Invalid resource requirements.  Resource requirements unchanged.")); /* catgets  2116  */
      clrtoeol();
   } else {
      
      move(1,0);
      strcpy(resreq, newresreq);
   } 
   clrtoeol();
   refresh();
}

int
main(int argc, char **argv)
{

    char   *cmd;             
    struct stat statBuf; 
    fd_set mask, allMask;    
    int    nReady;           
    int StdIn = 0;           
    int	optc;

    
    
    int Vflag = 0, Nflag = 0, Eflag = 0, Rflag = 0, iflag = 0, nflag = 0, Iflag = 0;
    extern int optind, opterr;
    extern char *optarg;
    int rc;

    Signal_(SIGINT, (SIGFUNCTYPE) quit);

    
    numneeded = 0;
    cmd = *argv;                               

    opterr = 0;
    rc = _i18n_init ( I18N_CAT_MIN );	


    while ((optc = getopt(argc, argv, "hVNEn:R:I:i:L:")) != EOF)
    {
	switch(optc)
	{
	case 'R':
	    if (Rflag)
		usage(cmd);
	    else {
		Rflag = 1;
		strcpy(resreq , optarg);
		break;
	    }
        case 'I':
            if (Iflag)
		usage(cmd);
	    else {
		Iflag = 1;
		strcpy(indexfilter, optarg);
		break;
	    }

	case 'n':		
	    if (nflag)
		usage(cmd);
	    else {
		nflag = 1;
		numneeded = atoi(optarg);
		if (numneeded <= 0) {
		    usage(cmd);
		}
		break;
	    }

	case 'i':
	    if (iflag)
		usage(cmd);
	    else {
		iflag = 1;
		updateRate = atoi(optarg);
		if (updateRate <= 0) {
		    usage(cmd);
		}
		break;
	    }
	    
        case 'L':		
	    if (Lflag)		
		usage(cmd);
	    else {
		Lflag = 1;
		strcpy(logFilename, optarg);
		break;
	    }
	    
	case 'N':
	    if (Nflag || Eflag)
		usage(cmd);
	    else {
		Nflag = 1;
		options = NORMALIZE;
		break;
	    }

	case 'E':
	    if (Nflag || Eflag)
		usage(cmd);
	    else {
		Eflag = 1;
		options = EFFECTIVE;
		break;
	    }

	case 'V':
	    if (Vflag)
		usage(cmd);
	    else {
		Vflag = 1;
		fputs(_LAVALITE_VERSION_, stderr);
		exit(0);
	    }

	case 'h':
	default:
            usage(cmd);
	}

    }


    while (optind < argc)		
    {
       hostnames[num] = argv[optind++];
       if (ls_isclustername(hostnames[num]) <=0 &&
          !isValidHost_(hostnames[num])) {  
	  fputs(_i18n_msg_get(ls_catd,NL_SETN,2141, "Unknown host: "), /* catgets  2141  */
	      stderr); 
	  fputs(hostnames[num], stderr);
	  putc('\n', stderr);
	  usage(cmd);
	  }
       num++;
       if (num>=MAXLISTSIZE) {
	  fprintf(stderr, _i18n_msg_get(ls_catd,NL_SETN,2142, "too many hosts specified\n")); /* catgets  2142  */
	  exit(-1);
       }
    }
  
    if (Lflag) {
        lf = fopen(logFilename, "a");
	if (lf == NULL) {
	    fprintf(stderr, I18N_FUNC_FAIL_S, logFilename, "fopen", strerror(errno));
	    fprintf(stderr, "\n" );
    	    exit(-1);
	}
    }

    if (fstat(0, &statBuf) == 0) 
	redirect = ((ftell(stdin) == 0) && Lflag) ? 1 : 0;

    if (!redirect) {
        
    (void)initscr();
        cbreak(); 
        noecho();

        
        if ((COLS < 80)) {
            (void)endwin();
            (void)fprintf(stderr,
	              _i18n_msg_get(ls_catd,NL_SETN,2144, "Sorry, screen must be at least %d COLUMNS wide (currently COLS = %d).\n"), /* catgets  2144  */
		      80, COLS); 
            exit(1);
        }

        if (LINES < 5) {
            (void)endwin();
            (void)fprintf(stderr,
		      _i18n_msg_get(ls_catd,NL_SETN,2145, "Sorry, screen must have at least %d LINES (currently LINES = %d).\n"), /* catgets  2145  */
		      5, LINES); 
            exit(1);
        } else if ((LINES < num + 4)) {
            
            (void)endwin();
            (void)fprintf(stderr,
                      _i18n_msg_get(ls_catd,NL_SETN,2146, "Sorry, screen must have at least %d LINES to display all of the requested hosts.\n"),  /* catgets  2146  */
		      num+4);
            (void)fprintf(stderr,
		      _i18n_msg_get(ls_catd,NL_SETN,2147, "The current window has %d lines.\n"), /* catgets  2147  */
		      LINES); 
            exit(1);
        }
        
        FD_ZERO(&allMask);
        FD_SET(StdIn, &allMask);
    } 


        
    for (;;) { 
	struct timeval timeout;

        getLoadInfo();

        if (!redirect)
        {
       	 	
  	    	timeout.tv_usec = 0;
	    	timeout.tv_sec = updateRate;

        	mask = allMask;
        	nReady = select(StdIn+1, &mask, 0, 0, &timeout);

	    	
	    	move(1, 0);
	    	clrtoeol();

                if ((nReady < 0) && (errno != EINTR)) {
		    char i18nBuf[100];
		    sprintf(i18nBuf, I18N_ERROR,"lsmon","select");
                    endPgm(i18nBuf, 1);
                } else if (nReady > 0) {
                    if (FD_ISSET(StdIn, &mask)) {
                     
                     switch (getch()) {
                     case CONTROL_L:
                         
                         erase();
                         refresh();
                         break;

                      case 'q':
                          
                          endPgm(NULL, 0);
                          break;
              
                      case 'i':
                          
                          chInterval();
                          break;
              
                      case 'n':
                          
                          numneeded = chNumNeeded();
                          break;
              
                      case 'N':
                          
                          move(1,0);
                          clrtoeol();
                          if (options == NORMALIZE) {
                              options = 0;
                              mvaddstr(1, 0,
                                _i18n_msg_get(ls_catd,NL_SETN,2149, "Output raw CPU run queue load indices")); /* catgets  2149  */
                          } else {
                              options = NORMALIZE;
                              mvaddstr(1, 0, 
			       	_i18n_msg_get(ls_catd,NL_SETN,2150, "Output normalized CPU run queue load indices")); /* catgets  2150  */
                          }
                          move(1,0);
                          refresh();
                          break;
              
                      case 'E':
                          
                          move(1,0);
                          clrtoeol();
                          if (options == EFFECTIVE) {
                              options = 0;
                              mvaddstr(1, 0,
                                  _i18n_msg_get(ls_catd,NL_SETN,2151, "Output raw CPU run queue load indices")); /* catgets  2151  */
                          } else {
                              options = EFFECTIVE;
                              mvaddstr(1, 0, 
			        _i18n_msg_get(ls_catd,NL_SETN,2152, "Output effective CPU run queue load indices")); /* catgets  2152  */
                          }
                          move(1,0);
                          refresh();
                          break;

                      case 'R':
                          
                          chResReq();
                          break;

                      default:
                          
                          move(1,0);
                          clrtoeol();
                          mvaddstr(1, 0, 
			    _i18n_msg_get(ls_catd,NL_SETN,2153, "Options:(i)nterval,(n)umber,(N)ormalize,(E)ffective,(R)esources,(q)uit,(^L)refresh")); /* catgets  2153  */
                          move(1,0);
                          refresh();
                          break;
                      }
                    }   
                }   

        }   
        else 
            sleep(updateRate);
    }             

    return 0;
}

